### ISSUE TITLE
[BUG/ISSUE] Error while compiling GCHP version 13

### ISSUE THREAD

**Issue by mtsivlidou** [geoschem/GCHP#80]
Hello, 
I am trying to compile the latest version of GCHP, with intel openmpi. I follow the steps according to the Youtube tutorial and the documentation, however I get an error after configuring my build, when I use "make -j". 

In build directory I use:
 cmake ..        (log file: [log.cmake.txt](https://github.com/geoschem/GCHP/files/5876499/log.cmake.txt)

 cmake .        (log file: [log.cmake2.txt](https://github.com/geoschem/GCHP/files/5876502/log.cmake2.txt))

 make -j        (log file: [log.make.txt](https://github.com/geoschem/GCHP/files/5876503/log.make.txt))

The error I get is:
/scratch/cnt0022/lat0005/tsivlidou/Code.GCHP/src/yaFyaml/src/Parser.F90(106): error #7497: The variable must be type compatible with the expression and of the same rank.   [TOKEN]
       token = lexr%get_token()

### GEOS-Chem version
GCHP version 13

### Software versions
- Compilers : intel/18.1
- NetCDF version: hdf5/1.8.18 netcdf/4.4.0_fortran-4.4.2
- MPI version: openmpi/intel/2.0.4
- ESMF version: 8.0.0

I would appreciate it if you could provide some help. 

Thank you in advance, 
Maria Tsivlidou

**Comment by LiamBindle** [comment]
Hi @mtsivlidou, thanks for opening this issue. I opened https://github.com/Goddard-Fortran-Ecosystem/yaFyaml/issues/27. I suspect this might be a compiler incompatibility. I believe GMAO plans to support the latest minor release of Intel 18 compilers, but I'm not sure how well tested it is. Unfortunately, it's hard for us to cover Intel compilers in our automated build tests because they aren't free and threfore they are difficult to include.

If it's possible to upgrade to a newer Intel or GNU compilers, I suspect that would fix the issue.

**Comment by yantosca** [comment]
Hi @mtsivlidou.  I ran up against a similar problem in that the GMAO libraries skew towards supporting the most recent compiler versions, because they take advantage some of the newest features in the newest Fortran standards (like F2018).  I was using Gfortran 9.3.0 and that still wasn't "new" enough even though that's a pretty recent release.

**Comment by yantosca** [comment]
Although I think GMAO put a patch in for gfortran 9.3.0.  We can compile GCHP locally at Harvard with 

- gcc/gfortran 9.3.0
- OpenMPI 4.0.2
- netCDF-Fortran 4.5.2
- netCDF-C 4.7.3
- HDF5 1.10.6

**Comment by LiamBindle** [comment]
We are using GNU 8.3.1 on Compute1 without any issues. I think we should treat any compiler errors with the latest minor version of Intel 18 or 19 and GNU 8 or 9 a bugs. 

@mtsivlidou is your compiler Intel 18.0.1? I don't see an 18.1 release on the Intel website. Perhaps I'm looking in the wrong place?

One other thing (unrelated to this issue, but it might crop up later). I see your MPI version is listed as `openmpi/intel/2.0.4`. Is that OpenMPI 2.0.4? (you can check with `mpirun --version`). If so, that might be too old. I think you will need OpenMPI >= 3.0.

**Comment by mtsivlidou** [comment]
Hello, 

Thank you for your replies. 
The compiler is intel 18.1 and the MPI version is Open MPI 2.0.4 (I was able to compile GCHP v 12.9.3 using this MPI version, does this mean that it might be enough for GCHP v.13 as well?). 
In the cluster I work there is intel 19.3 and 19.4 but not a higher version of Open MPI. I will try to compile again with intel 19.4 to see if it makes any difference. 
Maria

**Comment by yantosca** [comment]
Thanks for the update @mtsivlidou.  We haven't done a broad spectrum testing with different MPI versions.  If you do ever get the chance to use an updated MPI, we would recommend OpenMPI 4 or higher.  (OpenMPI3 had problems that I think were resolved in OpenMPI4).  There's also IntelMPI if you have that available.

Also -- GCST member @WilliamDowns has been able to port GCHP to Spack (hot off the presses!).  I think Will is currently working on the documentation for this.  The idea is that you clone Spack from their github site and do "spack install gchp" (plus some customization options). This willinstall all required libraries, plus the GCHP source code and executable.   I would defer to Will for more instructions about using this feature (as the doc is not 100% ready).  If you do want to use this, it might be a good test case for us.

**Comment by WilliamDowns** [comment]
Hi @mtsivlidou, as Bob said GCHP and its dependencies are now installable through Spack. While the exact documentation is not yet online, you can try the following steps to build GCHP with a new compatible version of OpenMPI (as well as new copies of other libraries):

1. Clone the latest version of Spack by typing `git clone https://github.com/spack/spack.git`
2. Execute the following code: `export SPACK_ROOT=/path/to/spack; . $SPACK_ROOT/share/spack/setup-env.sh`, replacing "`/path/to/spack`" with the location of your Spack repo. This path will be where all installed packages are stored. You should save this code to an environment file as you will need to execute these commands to use Spack whenever you login to your cluster.
3. Load a version of your Intel compiler however you normally would.
4. Type `spack compiler find`, which will cause Spack to recognize your Intel compiler so you don't need to load it independently of Spack in the future.
5. If you plan to use Slurm to submit GCHP jobs, create a new file `$HOME/.spack/packages.yaml` and insert the following lines: 

```
packages:
 slurm:
  buildable: false
  externals:
  - spec: slurm
    prefix: /path/to/slurm
```
Where `/path/to/slurm` is the location of your cluster's Slurm installation. I do not know a method of automatically locating this directory at this time. On our cluster the path is `/usr/include/slurm`. 
6. Type `spack install gchp%intel ^openmpi@4.0.5 +pmi schedulers=slurm` if you will be using Slurm, or `spack install gchp%intel ^openmpi@4.0.5` if you will not be using Slurm. This will install new Spack-built versions of the libraries required for GCHP, including OpenMPI 4.0.5, and then build GCHP itself.
7. Once this is finished (it may take awhile because Spack will install numerous libraries), you will have a built `gchp` executable and a copy of the GCHP source code which you can locate using `spack location -i gchp`. You can create run directories from here.
8. When running GCHP, use `spack load gchp` to load all of GCHP's required libraries.

**Comment by LiamBindle** [comment]
Hi @mtsivlidou, are you still encountering problems? It sounds like this might be a compiler incompatibility. Is upgrading to a newer compiler possible (at least Intel 18 Update 5, or GNU >= 8.3)?

**Comment by mtsivlidou** [comment]
Hello again, 

In order to face this error #116, I was adviced to compile the lastest version of the model. 
For this, I need to ask the support of the cluster I am working to install the suitable libraries for me.
Could you please confirm that the packages:
- gcc/gfortran 9.3.0
- OpenMPI 4.0.2
- netCDF-Fortran 4.5.2
- netCDF-C 4.7.3
- HDF5 1.10.6 

are suggested for the long-term runs with GCHP? 
Thank you in advance, Maria

**Comment by LiamBindle** [comment]
Hi Maria, those should be good.

If you run into a compiler error with these libraries, could you open a new issue? If you could share your CMakeCache.txt and the compiler log that should be all the info we need. GCC 9.3 should be okay (I'm using 8.3). I wouldn't expect there to be any issues with these versions, so if you do run into problems I'd like to make sure we fix them.

It would be good to use 13.1.2 too (or preferably, 13.2.1 if it's available when you start).

**Comment by mtsivlidou** [comment]
Hello, 
I will install 13.1.2 or 13.2.1 as soon as the installation of the libraries is finished. Yes sure, I will open a new issue in case I face any compiler error. Thank you!
Maria

**Comment by xpye98** [comment]
@mtsivlidou I encountered the same error with Intel compiler 18.0.0, and the solution [here](https://github.com/Goddard-Fortran-Ecosystem/yaFyaml/issues/27) can fix the problem well, at least now I can build successfully.
